### PR TITLE
API: Add and redefine `numpy.bool` [Array API]

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @rgommers @ngoldbaum @seberg,

One of the members of Array API is [bool](https://data-apis.org/array-api/latest/API_specification/data_types.html?highlight=bool#bool). Currently `np.bool_` is present and I think we already had some initial conversations in NEP 52 issues about renaming it to `np.bool`.

Would you like to keep both names in the long term? Or should I remove `np.bool_` usages and deprecate it?

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Some context from prior discussions: https://github.com/numpy/numpy/issues/24306#issuecomment-1664042728

The deprecation happened in 1.20 (Jan 2021) and was finalized in NumPy 1.24 (Dec 2022).

Is a year enough time for projects to migrate? I'm honestly not sure. Some light github code searches indicate that usages of "np.bool" are mostly in code that hasn't been touched in several years.

**Comment by rgommers** [comment]
I don't see any problem with putting `np.bool` back. We should never have even removed it I believe; IIRC we discussed restoring it immediately because the deprecation was not a good choice. We just failed to follow through on that.

Either way, a couple more arguments:
- 2.0 is a major release, so a good time for this
- This is an essential part of array API support, which we already decided we wanted to do back in the 2.0 developer meeting in April (NEP to expand on that to follow)
- Code that hasn't been touched in several years isn't really that interesting, it's probably still not going to be touched in 1 or 2 years from now
- The linked discussion in [gh-24306](https://github.com/numpy/numpy/issues/24306#issuecomment-1664042728) is mostly about the `True_` and `False_` singletons, which is related but not the same thing. And 99.x% of usage is through `dtype=np.bool` keywords, which is compatible with the old usage. Calling `np.bool(a_value)` is very niche in comparison.

**Comment by rgommers** [comment]
> Or should I remove `np.bool_` usages and deprecate it?

We should keep that for quite a while, in order to not force people to write if/else stuff depending on the numpy version. In the long run it can go, but there is no hurry I think.

**Comment by ngoldbaum** [comment]
> Code that hasn't been touched in several years isn't really that interesting, it's probably still not going to be touched in 1 or 2 years from now

Agreed, I was trying to say that the migration appears mostly complete in downstream code.

I think a year is probably long enough given Ralf's points.

**Review by BvB93 — COMMENTED** [review]
The main `numpy/__init__.pyi` file also needs a `bool = bool_` declaration for type checkers to pick up the alias. 

This is actually gona be somewhat cumbersome however, as this would result in a naming conflict with the builtin `bool` type that can only be resolved via replacing all previous mentions of `bool` in the module with an explicit `builtins.bool`.

For reference: we've had similar issues before inside the `dtype` class due to a `dtype.str`/`builtins.str` naming conflict (xref https://github.com/numpy/numpy/pull/20224)

**Comment by seberg** [comment]
If we reintroduce `bool` the name should maybe be the main name and `bool_` the (hidden?) alias?  (this only really changes how `np.bool_` is printed)
There is also the open question if people prefer `np.bool(True)` over `np.True_`.

**Comment by rgommers** [comment]
Good point, +1 for making `np.bool` the main/preferred name.

> There is also the open question if people prefer `np.bool(True)` over `np.True_`.

Slight preference for `np.bool(True)` here, seems nicer - but also happy if others feel differently.

**Comment by ngoldbaum** [comment]
Another +1 for `np.bool(True)`. It also moves us more in the direction of eventually getting rid of `np.True_` (which we decided was too disruptive for 2.0, but maybe not for 3.0...) instead of permanently enshrining it in the API.

**Comment by seberg** [comment]
One thing about `np.True_`/`np.False_` is that they are true singletons and having them makes more obvious, but no big opinion.

**Comment by eric-wieser** [comment]
> I don't see any problem with putting `np.bool` back. We should never have even removed it I believe; IIRC we discussed restoring it immediately because the deprecation was not a good choice. We just failed to follow through on that.


I'm a bit confused by this claim; this PR is *not* putting `np.bool` back; it's replacing `np.bool` with something different to what it previously contained, and the only sensible way to achieve that was to remove the old version first via a deprecation.

Perhaps the PR title and release note should be clear that this is not a "reintroduction", but a reallocation of the name to a new thing.

File: doc/release/upcoming_changes/25080.change.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+* Redefine `numpy.bool` as an alias for `numpy.bool_` for Array API compatibility.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/25080.change.rst:1]:
```suggestion
* Redefine `numpy.bool` as an alias for `numpy.bool_` (as opposed to the `bool` it was in NumPy 1.x) for Array API compatibility.
```

**Comment by rgommers** [comment]
> I'm a bit confused by this claim; this PR is _not_ putting `np.bool` back; it's replacing `np.bool` with something different to (and better than) what it previously contained, and the only sensible way to achieve that was to remove the old version first via a deprecation.

Yes I understand that, and no it wasn't the only way to go about this. We now effectively made everyone change their code twice, while the vast majority of usage of the old `np.bool`, which are `func(..., dtype=np.bool`, would have worked fine when making the change to `np.bool is np.bool_` without a deprecation. As would `if np.bool(a_value) > 10:` type things. In effect, we caused a lot of extra churn for limited benefit. It's a trade-off; making users change code twice is at least a "deprecation smell" though.

I am pretty sure we discussed undoing the `np.bool` deprecation after it was done, and concluded that we should probably undo that before the then-next release. We just then lost track of that, and never revisited the topic.

File: doc/release/upcoming_changes/25080.change.rst (hunk 1 above)
Comment (mtsokol) [reply doc/release/upcoming_changes/25080.change.rst:1]: Sure! Updated.

**Comment by mtsokol** [comment]
@rgommers @seberg I started working on making `np.bool` default one and `np.bool_` a hidden alias (this will require ~1200 lines of code changed, mostly in typing to use `np.bool` - all of them simple changes).

I already pushed ~900 of them, I still need to fix a few `pyi` files - let me know if this scope of changes is fine/too disruptive for you.

**Comment by rgommers** [comment]
If they're all simple replacements then I think it's fine to do now - but it'd be completely fine to not do it as well. In the end what matters is what we write in the docs I think (and the repr etc. - user-visible stuff); the `bool_` -> `bool` replacements can happen later too.

**Comment by seberg** [comment]
I honestly don't care if we do it now.  I was mostly thinking of what the `repr()` shows.

**Comment by mtsokol** [comment]
Ok, now reprs for False and True are updated. The PR is ready for a review.

File: doc/release/upcoming_changes/25080.change.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+* Redefine `numpy.bool` as an alias for `numpy.bool_` (as opposed to the `bool`
+it was until NumPy 1.24) for Array API compatibility.
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/25080.change.rst:2]: Maybe also make a separate release note for the repr of `np.True_` and `np.False_`, mentioning that those are still available but the canonical names are now `np.bool(True)` and `np.bool(False)`.

Comment (eric-wieser) [reply doc/release/upcoming_changes/25080.change.rst:2]: Aren't those canonical names slower?

Comment (mtsokol) [reply doc/release/upcoming_changes/25080.change.rst:2]: Resolving as we decided to stay with `np.True_` and `np.False_` for reprs.

File: numpy/_core/defchararray.pyi [hunk 3]
```diff
@@ -27,7 +28,7 @@ from numpy._typing import (
     _ArrayLikeStr_co as U_co,
     _ArrayLikeBytes_co as S_co,
     _ArrayLikeInt_co as i_co,
-    _ArrayLikeBool_co as b_co,
+    _ArrayLikebool_co as b_co,
```
Comment (rgommers) [inline numpy/_core/defchararray.pyi:31]: This change looks incorrect, it's still `_ArrayLikeBool_co` in `numpy._typing`.

File: numpy/_core/fromnumeric.pyi [hunk 4]
```diff
@@ -8,7 +9,7 @@ from numpy import (
     int64,
     intp,
     float16,
-    bool_,
+    bool,
```
Comment (rgommers) [inline numpy/_core/fromnumeric.pyi:12]: The diff would be a lot smaller if you'd leave out this import. Using the builtin can then stay unchanged. Seems preferable to me to use `np.bool` (unless that doesn't work?).

File: numpy/_core/numeric.py [hunk 5]
```diff
@@ -2531,13 +2530,13 @@ def array_equiv(a1, a2):
...
-    return bool((a1 == a2).all())
+    return builtins.bool((a1 == a2).all())
 
 
 inf = PINF
 nan = NAN
-False_ = bool_(False)
-True_ = bool_(True)
+False_ = nt.bool(False)
+True_ = nt.bool(True)
```
Comment (rgommers) [inline numpy/_core/numeric.py:2539]: Using `np.bool` would be clearer; this relies on a weird replacement thing in `numerictypes.py` - it does a `from builtins import bool` and then overwrites that with:
```python
# Now add the types we've determined to this module
for key in allTypes:
    globals()[key] = allTypes[key]
    __all__.append(key)
```

**Review by rgommers — COMMENTED** [review]
This is looking quite good. Everything works as far as I can tell. I have a couple of small comments inline. My main one is about the type stubs, where you replaced `from numpy import _bool` with `from numpy import bool`. And as a result have a lot of `bool` -> `builtins.bool` changes. Those may in some cases be useful anyway for clarity, however wouldn't it be better to _not_ do `from numpy import bool`?

Then you'd have to write
```python
NDArray[np.bool]
```
instead of
```python
NDArray[bool]
```
but that seems useful anyway.

File: numpy/_core/numeric.py (hunk 5 above)
Comment (mtsokol) [reply numpy/_core/numeric.py:2539]: I noticed it, but with `np.bool` importing NumPy fails with:

```python
AttributeError: partially initialized module 'numpy' has no attribute 'bool' (most likely due to a circular import)
```
as `numpy.core` module is imported eagerly (`__init__.py` imports `core/numeric.py` which imports `__init__.py`).

I would leave `nt.bool`, or import it locally with a different name? WDYT?

Comment (rgommers) [reply numpy/_core/numeric.py:2539]: Ah okay. In that case, maybe leave it as is, but ensure that there's a test that `_core.numerictypes.bool is np.bool` to guard against a future issue there?

**Comment by rgommers** [comment]
> Sure - I can use `np.bool` for NumPy's bool and `bool` for builtin, but I will update it this way in **all** stub files, so it's consistent across the codebase.

Seems like a good idea to me, but let's ask @BvB93 if he has any preference for these stub files.

File: numpy/_core/defchararray.pyi (hunk 3 above)
Comment (mtsokol) [reply numpy/_core/defchararray.pyi]: Right - fixed it.

File: numpy/_core/numeric.py (hunk 5 above)
Comment (mtsokol) [reply numpy/_core/numeric.py:2523]: Sure! I added a test for that.

**Comment by mtsokol** [comment]
> > Sure - I can use `np.bool` for NumPy's bool and `bool` for builtin, but I will update it this way in **all** stub files, so it's consistent across the codebase.
> 
> Seems like a good idea to me, but let's ask @BvB93 if he has any preference for these stub files.

@rgommers I updated the PR - now NumPy's bool is accessed as `np.bool` only, and builtin bool as `bool`.

Could we merge this PR? If we come up with another preference/arrangement for stub files I will adjust it in a separate one.

**Comment by rgommers** [comment]
The one CI failure is `RuntimeWarning: invalid value encountered in matmul`, at first sight that looks unrelated.

**Comment by seberg** [comment]
> The one CI failure is RuntimeWarning: invalid value encountered in matmul, at first sight that looks unrelated

That has been a heisenbug (probably in openblas) since forever.

**Review by rgommers — APPROVED** [review]
> Could we merge this PR? If we come up with another preference/arrangement for stub files I will adjust it in a separate one.

Agreed. This all looks solid, so let's go with it. Thanks Mateusz & reviewers!

**Pull request merged**